sdk/search/azure-search-documents/src/main/java/com/azure/search/documents/indexes/SearchIndexAsyncClient.java

```diff
@@ -0,0 +1,369 @@
+import com.azure.core.http.rest.PagedFlux;
+import com.azure.core.http.rest.PagedResponse;
+import com.azure.core.http.rest.PagedResponseBase;
+import com.azure.core.http.rest.Response;
+import com.azure.core.util.Context;
+import com.azure.core.util.FluxUtil;
+import com.azure.core.util.logging.ClientLogger;
+import com.azure.search.documents.SearchServiceVersion;
+import com.azure.search.documents.implementation.SearchServiceRestClientBuilder;
+import com.azure.search.documents.implementation.SearchServiceRestClientImpl;
+import com.azure.search.documents.models.AnalyzeRequest;
+import com.azure.search.documents.models.AnalyzedTokenInfo;
+import com.azure.search.documents.models.GetIndexStatisticsResult;
+import com.azure.search.documents.models.RequestOptions;
+import com.azure.search.documents.models.SearchIndex;
+import reactor.core.publisher.Mono;
+
+import java.util.Objects;
+import java.util.function.Function;
+
+import static com.azure.core.util.FluxUtil.monoError;
+import static com.azure.core.util.FluxUtil.pagedFluxError;
+import static com.azure.core.util.FluxUtil.withContext;
+
+/**
+ * Asynchronous Client to manage and query search index, as well as manage other resources,
+ * on a Cognitive Search service.
+ */
+public class SearchIndexAsyncClient {
+    /**
+     * The logger to be used
+     */
+    private final ClientLogger logger = new ClientLogger(SearchIndexAsyncClient.class);
+
+    /**
+     * The underlying AutoRest client used to interact with the Search service
+     */
+    private final SearchServiceRestClientImpl restClient;
+
+    SearchIndexAsyncClient(String endpoint, SearchServiceVersion serviceVersion, HttpPipeline httpPipeline) {
+        this.restClient = new SearchServiceRestClientBuilder()
+            .endpoint(endpoint)
+            .apiVersion(serviceVersion.getVersion())
+            .pipeline(httpPipeline)
+            .build();
+    }
+
+    /**
+     * Creates a new Azure Cognitive Search index.
+     *
+     * @param index definition of the index to create.
+     * @return the created Index.
+     */
+    public Mono<SearchIndex> create(SearchIndex index) {
+        return createWithResponse(index, null).map(Response::getValue);
+    }
+
+    /**
+     * Creates a new Azure Cognitive Search index.
+     *
+     * @param index definition of the index to create
+     * @param requestOptions additional parameters for the operation. Contains the tracking ID sent with the request to
+     * help with debugging
+     * @return a response containing the created Index.
+     */
+    public Mono<Response<SearchIndex>> createWithResponse(SearchIndex index, RequestOptions requestOptions) {
+        return withContext(context -> createWithResponse(index, requestOptions, context));
+    }
+
+    Mono<Response<SearchIndex>> createWithResponse(SearchIndex index, RequestOptions requestOptions,
+        Context context) {
+        Objects.requireNonNull(index, "'Index' cannot be null");
```

Comment:
We should continue this with monoError(logger, new NullPointerException("'index' cannot be null."); This will throw rhater than return that mono error we want.

Reply:
Good catch. Will rework on all occurrences.

Reply:
Or just include it within the `try {} catch()` so that it is raised as a monoError

Reply:
Not sure if this applies to Java, but try/catch is expensive in .NET where we can just throw the exception as-is.